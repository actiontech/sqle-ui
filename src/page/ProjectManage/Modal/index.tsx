import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { ModalName } from '../../../data/ModalName';
import { initProjectManageModalStatus } from '../../../store/projectManage';
import CreateProject from './CreateProject';
import UpdateProject from './UpdateProject';

const ProjectManageModal: React.FC = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(
      initProjectManageModalStatus({
        modalStatus: {
          [ModalName.Create_Project]: false,
          [ModalName.Update_Project]: false,
        },
      })
    );
  }, [dispatch]);

  return (
    <>
      <CreateProject />
      <UpdateProject />
    </>
  );
};

export default ProjectManageModal;
