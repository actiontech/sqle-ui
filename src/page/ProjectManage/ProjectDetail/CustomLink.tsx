import { Link } from 'react-router-dom';
import { CustomLinkProps, ProjectDetailCustomLinkState } from '.';

/**
 *
 * @param param0
 * @returns
 */
const CustomLink: React.FC<CustomLinkProps> = ({
  to,
  projectName,
  search,
  children,
  ...props
}) => {
  return (
    <Link<ProjectDetailCustomLinkState>
      {...props}
      to={{ pathname: to, search, state: { projectName } }}
    >
      {children}
    </Link>
  );
};

export default CustomLink;
