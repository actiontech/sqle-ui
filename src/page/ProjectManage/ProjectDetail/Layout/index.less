.project-detail-wrapper {
  min-height: calc(100vh - 50px);

  .project-info-wrapper {
    padding: 20px;
    .title {
      margin-bottom: 16px;
      font-size: 24px;
    }
    .pb-10 {
      padding-bottom: 10px;
    }
  }
}
