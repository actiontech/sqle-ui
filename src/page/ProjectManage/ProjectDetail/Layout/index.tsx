import { Menu, Tooltip } from 'antd';
import { ReactNode } from 'react';
import { Link } from 'react-router-dom';
import { ProjectDetailUrlParamType } from '..';
import { IProjectDetailItem } from '../../../../api/common';
import { SystemRole } from '../../../../data/common';
import i18n from '../../../../locale';
import {
  ProjectDetailRouterItemKeyLiteral,
  RouterItem,
} from '../../../../types/router.type';
import ProjectDetailLayout from './Layout';

export type ProjectDetailLayoutProps = {
  children: ReactNode;
  archive: boolean;
} & ProjectDetailUrlParamType;

export type ProjectInfoBoxProps = {
  projectInfo?: IProjectDetailItem;
};

export const generateNavigateMenu = (
  config: Array<RouterItem<ProjectDetailRouterItemKeyLiteral>>,
  userRole: SystemRole | '',
  projectName: string
) => {
  return config.map((route) => {
    if (Array.isArray(route.role) && !route.role?.includes(userRole)) {
      return null;
    }
    if (route.hideInSliderMenu) {
      return null;
    }
    if (!!route.components) {
      return (
        <Menu.SubMenu
          key={route.key}
          icon={route.icon}
          title={route.labelWithoutI18n ?? i18n.t(route.label)}
        >
          {generateNavigateMenu(route.components, userRole, projectName)}
        </Menu.SubMenu>
      );
    }

    if (!!route.groups) {
      return route.groups.map((v) => {
        return (
          <Menu.ItemGroup title={v.title} key={v.title}>
            {generateNavigateMenu(v.values, userRole, projectName)}
          </Menu.ItemGroup>
        );
      });
    }

    return (
      <Menu.Item
        key={route.key}
        icon={route.icon}
        title={route.labelWithoutI18n ?? i18n.t(route.label)}
      >
        <Link to={(route.path as string).replace(':projectName', projectName)}>
          {route.labelWithoutI18n ?? i18n.t(route.label)}
        </Link>
      </Menu.Item>
    );
  });
};

export default ProjectDetailLayout;
