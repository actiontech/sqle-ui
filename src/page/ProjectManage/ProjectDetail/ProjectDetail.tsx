import { useRequest } from 'ahooks';
import { Result } from 'antd';
import { Suspense, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';
import { Outlet } from 'react-router-dom';
import { useCurrentProjectName } from '.';
import project from '../../../api/project';
import HeaderProgress from '../../../components/HeaderProgress';
import useUserInfo from '../../../hooks/useUserInfo';
import { updateProjectStatus } from '../../../store/projectManage';
import ProjectDetailLayout from './Layout';

const ProjectDetail: React.FC = () => {
  const { projectName } = useCurrentProjectName();
  const { t } = useTranslation();
  const dispatch = useDispatch();

  const { getUserInfo, getUserInfoLoading, userInfo } = useUserInfo();

  const { data: projectInfo, loading: getProjectInfoLoading } = useRequest(
    () =>
      project.getProjectDetailV1({ project_name: projectName }).then((res) => {
        dispatch(updateProjectStatus(!!res.data.data?.archived));
        return res.data.data;
      }),
    {
      ready: !!projectName,
      refreshDeps: [projectName],
    }
  );

  useEffect(() => {
    getUserInfo();
  }, [getUserInfo]);

  const renderProjectDetail = () => {
    if (getUserInfoLoading || getProjectInfoLoading) {
      return <HeaderProgress />;
    }
    if ((userInfo?.data.data?.bind_projects?.length ?? 0) === 0) {
      return (
        <Result
          status="info"
          title={t('projectManage.projectDetail.notice')}
          subTitle={t('projectManage.projectDetail.unboundProjectTips')}
        />
      );
    }

    return (
      <ProjectDetailLayout
        projectName={projectName}
        archive={!!projectInfo?.archived}
      >
        <Suspense fallback={<HeaderProgress />}>
          <Outlet />
        </Suspense>
      </ProjectDetailLayout>
    );
  };

  return <>{renderProjectDetail()}</>;
};

export default ProjectDetail;
