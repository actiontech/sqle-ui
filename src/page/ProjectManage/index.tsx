import { PageHeader } from 'antd';
import { useTranslation } from 'react-i18next';
import ProjectManageModal from './Modal';
import ProjectList from './ProjectList';

const ProjectManage: React.FC = () => {
  const { t } = useTranslation();
  return (
    <article className="project-manage-page-namespace">
      <PageHeader title={t('projectManage.pageTitle')} ghost={false}>
        {t('projectManage.pageDescribe')}
      </PageHeader>

      <ProjectList />
      <ProjectManageModal />
    </article>
  );
};

export default ProjectManage;
