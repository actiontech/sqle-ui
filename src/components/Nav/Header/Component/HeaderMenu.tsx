import { useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { useHistory, useLocation } from 'react-router-dom';
import { DEFAULT_PROJECT_NAME } from '../../../../page/ProjectManage/ProjectDetail';
import { globalRouterConfig } from '../../../../router/config';
import {
  GlobalRouterItemKeyLiteral,
  RouterItem,
} from '../../../../types/router.type';

const headerMenuKeys: Array<typeof globalRouterConfig[number]['key']> = [
  'dashboard',
  'rule',
  'sqlQuery',
  /* IFTRUE_isEE */
  'projectList',
  /* FITRUE_isEE */
  /* IFTRUE_isCE */
  'projectDetail',
  /* FITRUE_isCE */
];

const HeaderMenu: React.FC = () => {
  const { t } = useTranslation();
  const location = useLocation();
  const history = useHistory();

  const jumpToPath = (path: string) => {
    /* IFTRUE_isCE */
    if (path.includes(':projectName')) {
      path = path.replace(':projectName', DEFAULT_PROJECT_NAME);
    }
    /* FITRUE_isCE */

    history.push(path);
  };

  const isActiveMenu = useCallback(
    (router: RouterItem<GlobalRouterItemKeyLiteral>) => {
      if (router.key === 'projectList' || router.key === 'projectDetail') {
        return /^\/project\/.+/.test(location.pathname);
      }

      return router.path === location.pathname;
    },
    [location.pathname]
  );
  return (
    <div className="header-menu">
      {globalRouterConfig.map((router) => {
        if (headerMenuKeys.includes(router.key)) {
          return (
            <div
              key={router.key}
              className={`${
                isActiveMenu(router) ? 'header-menu-item-active' : ''
              } header-menu-item`}
              onClick={() => jumpToPath(router.path as string)}
            >
              <span className="header-menu-item-icon">{router.icon}</span>
              {t(router.label)}
            </div>
          );
        }
        return null;
      })}
    </div>
  );
};

export default HeaderMenu;
