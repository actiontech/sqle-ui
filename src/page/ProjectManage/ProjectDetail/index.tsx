import ProjectDetail from './ProjectDetail';

export type ProjectDetailLocationStateType = {
  projectName: string;
};

export default ProjectDetail;
