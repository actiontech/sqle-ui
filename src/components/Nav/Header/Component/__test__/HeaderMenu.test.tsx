import {
  renderWithRouter,
  renderWithServerRouter,
} from '../../../../../testUtils/customRender';
import HeaderMenu from '../HeaderMenu';
import { createMemoryHistory } from 'history';
import { fireEvent, screen } from '@testing-library/react';
import { DEFAULT_PROJECT_NAME } from '../../../../../page/ProjectManage/ProjectDetail';

describe('test Nav/Header/HeaderMenu', () => {
  test('should match snapshot', () => {
    const { container } = renderWithRouter(<HeaderMenu />);

    expect(container).toMatchSnapshot();
  });

  test('should jump to path when clicking the menu item', () => {
    const history = createMemoryHistory();
    history.push('/test');
    renderWithServerRouter(<HeaderMenu />, undefined, { history });
    expect(history.location.pathname).toBe('/test');

    expect(screen.getByText('menu.dashboard')).not.toHaveClass(
      'header-menu-item-active'
    );
    fireEvent.click(screen.getByText('menu.dashboard'));
    expect(history.location.pathname).toBe('/');
    expect(screen.getByText('menu.dashboard')).toHaveClass(
      'header-menu-item-active'
    );

    expect(screen.getByText('menu.rule')).not.toHaveClass(
      'header-menu-item-active'
    );
    fireEvent.click(screen.getByText('menu.rule'));
    expect(history.location.pathname).toBe('/rule');
    expect(screen.getByText('menu.rule')).toHaveClass(
      'header-menu-item-active'
    );

    expect(screen.getByText('menu.sqlQuery')).not.toHaveClass(
      'header-menu-item-active'
    );
    fireEvent.click(screen.getByText('menu.sqlQuery'));
    expect(history.location.pathname).toBe('/sqlQuery');
    expect(screen.getByText('menu.sqlQuery')).toHaveClass(
      'header-menu-item-active'
    );

    expect(screen.getAllByText('menu.projectManage')[0]).not.toHaveClass(
      'header-menu-item-active'
    );
    fireEvent.click(screen.getAllByText('menu.projectManage')[0]);
    expect(history.location.pathname).toBe('/project');
    expect(screen.getAllByText('menu.projectManage')[0]).toHaveClass(
      'header-menu-item-active'
    );

    fireEvent.click(screen.getByText('menu.dashboard'));
    expect(screen.getAllByText('menu.projectManage')[1]).not.toHaveClass(
      'header-menu-item-active'
    );
    fireEvent.click(screen.getAllByText('menu.projectManage')[1]);
    expect(history.location.pathname).toBe(`/project/${DEFAULT_PROJECT_NAME}`);
    expect(screen.getAllByText('menu.projectManage')[1]).toHaveClass(
      'header-menu-item-active'
    );
  });
});
