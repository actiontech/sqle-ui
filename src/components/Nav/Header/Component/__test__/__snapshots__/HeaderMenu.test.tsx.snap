// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Nav/Header/HeaderMenu should match snapshot 1`] = `
<div>
  <div
    class="header-menu"
  >
    <div
      class=" header-menu-item"
    >
      <span
        class="header-menu-item-icon"
      >
        <span
          aria-label="pie-chart"
          class="anticon anticon-pie-chart"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="pie-chart"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M864 518H506V160c0-4.4-3.6-8-8-8h-26a398.46 398.46 0 00-282.8 117.1 398.19 398.19 0 00-85.7 127.1A397.61 397.61 0 0072 552a398.46 398.46 0 00117.1 282.8c36.7 36.7 79.5 65.6 127.1 85.7A397.61 397.61 0 00472 952a398.46 398.46 0 00282.8-117.1c36.7-36.7 65.6-79.5 85.7-127.1A397.61 397.61 0 00872 552v-26c0-4.4-3.6-8-8-8zM705.7 787.8A331.59 331.59 0 01470.4 884c-88.1-.4-170.9-34.9-233.2-97.2C174.5 724.1 140 640.7 140 552c0-88.7 34.5-172.1 97.2-234.8 54.6-54.6 124.9-87.9 200.8-95.5V586h364.3c-7.7 76.3-41.3 147-96.6 201.8zM952 462.4l-2.6-28.2c-8.5-92.1-49.4-179-115.2-244.6A399.4 399.4 0 00589 74.6L560.7 72c-4.7-.4-8.7 3.2-8.7 7.9V464c0 4.4 3.6 8 8 8l384-1c4.7 0 8.4-4 8-8.6zm-332.2-58.2V147.6a332.24 332.24 0 01166.4 89.8c45.7 45.6 77 103.6 90 166.1l-256.4.7z"
            />
          </svg>
        </span>
      </span>
      menu.dashboard
    </div>
    <div
      class=" header-menu-item"
    >
      <span
        class="header-menu-item-icon"
      >
        <span
          aria-label="project"
          class="anticon anticon-project"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="project"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M280 752h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v464c0 4.4 3.6 8 8 8zm192-280h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v184c0 4.4 3.6 8 8 8zm192 72h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v256c0 4.4 3.6 8 8 8zm216-432H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
            />
          </svg>
        </span>
      </span>
      menu.projectManage
    </div>
    <div
      class=" header-menu-item"
    >
      <span
        class="header-menu-item-icon"
      >
        <span
          aria-label="project"
          class="anticon anticon-project"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="project"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M280 752h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v464c0 4.4 3.6 8 8 8zm192-280h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v184c0 4.4 3.6 8 8 8zm192 72h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v256c0 4.4 3.6 8 8 8zm216-432H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
            />
          </svg>
        </span>
      </span>
      menu.projectManage
    </div>
    <div
      class=" header-menu-item"
    >
      <span
        class="header-menu-item-icon"
      >
        <span
          aria-label="search"
          class="anticon anticon-search"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="search"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
            />
          </svg>
        </span>
      </span>
      menu.sqlQuery
    </div>
    <div
      class=" header-menu-item"
    >
      <span
        class="header-menu-item-icon"
      >
        <span
          aria-label="desktop"
          class="anticon anticon-desktop"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="desktop"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M928 140H96c-17.7 0-32 14.3-32 32v496c0 17.7 14.3 32 32 32h380v112H304c-8.8 0-16 7.2-16 16v48c0 4.4 3.6 8 8 8h432c4.4 0 8-3.6 8-8v-48c0-8.8-7.2-16-16-16H548V700h380c17.7 0 32-14.3 32-32V172c0-17.7-14.3-32-32-32zm-40 488H136V212h752v416z"
            />
          </svg>
        </span>
      </span>
      menu.rule
    </div>
  </div>
</div>
`;
